File: Algorithms/1275. Find Winner on a Tic Tac Toe Game/golang/solution.go
func tictactoe(moves [][]int) string {
	rows := []int{0, 0, 0}
	cols := []int{0, 0, 0}

	diag := 0
	anti := 0

	remain := 9
	flag := true

	for i := 0; i < len(moves); i++ {
		if flag {
			rows[moves[i][0]]++
			cols[moves[i][1]]++
			if moves[i][0] == moves[i][1] {
				diag++
			}
			if moves[i][0]+moves[i][1] == 2 {
				anti++
			}
			flag = false
		} else {
			rows[moves[i][0]]--
			cols[moves[i][1]]--
			if moves[i][0] == moves[i][1] {
				diag--
			}
			if moves[i][0]+moves[i][1] == 2 {
				anti--
			}
			flag = true
		}

		if diag == 3 || anti == 3 || rows[0] == 3 || rows[1] == 3 || rows[2] == 3 || cols[0] == 3 || cols[1] == 3 || cols[2] == 3 {
			return "A"
		}
		if diag == -3 || anti == -3 || rows[0] == -3 || rows[1] == -3 || rows[2] == -3 || cols[0] == -3 || cols[1] == -3 || cols[2] == -3 {
			return "B"
		}
		remain--
	}

	if remain == 0 {
		return "Draw"
	}

	return "Pending"
}

File: Algorithms/1275. Find Winner on a Tic Tac Toe Game/python/solution.py
class Solution:
    def tictactoe(self, moves: list[list[int]]) -> str:

        # n stands for the size of the board, n = 3 for the current game.
        n = 3

        # use rows and cols to record the value on each row and each column.
        # diag1 and diag2 to record value on diagonal or anti-diagonal.
        rows, cols = [0] * n, [0] * n
        diag = anti_diag = 0

        # Two players having value of 1 and -1, player_1 with value = 1 places first.
        player = 1

        for row, col in moves:

            # Update the row value and column value.
            rows[row] += player
            cols[col] += player

            # If this move is placed on diagonal or anti-diagonal,
            # we shall update the relative value as well.
            if row == col:
                diag += player
            if row + col == n - 1:
                anti_diag += player

            # check if this move meets any of the winning conditions.
            if any(abs(line) == n for line in (rows[row], cols[col], diag, anti_diag)):
                return "A" if player == 1 else "B"

            # If no one wins so far, change to the other player alternatively.
            # That is from 1 to -1, from -1 to 1.
            player *= -1

        # If all moves are completed and there is still no result, we shall check if
        # the grid is full or not. If so, the game ends with draw, otherwise pending.
        return "Draw" if len(moves) == n * n else "Pending"
